sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
-                                          num_features)
-            left_source = current_node
+            pair_distance = dist_metric.dist(
+                &raw_data_view[current_node, 0],
+                &raw_data_view[j, 0],
+                num_features
+            )
 
             if alpha != 1.0:
-                left_value /= alpha
-
-            core_value = core_distances[j]
-            if (current_node_core_distance > right_value or
-                    core_value > right_value or
-                    left_value > right_value):
-                if right_value < new_distance:
-                    new_distance = right_value
-                    source_node = right_source
+                pair_distance /= alpha
+
+            next_node_core_dist = core_distances[j]
+            mutual_reachability_distance = max(
+                current_node_core_dist,
+                next_node_core_dist,
+                pair_distance
+            )
+            if mutual_reachability_distance > next_node_min_reach:
+                if next_node_min_reach < new_reachability:
+                    new_reachability = next_node_min_reach
+                    source_node = next_node_source
                     new_node = j
                 continue
 
-            if core_value > current_node_core_distance:
-                if core_value > left_value:
-                    left_value = core_value
-            else:
-                if current_node_core_distance > left_value:
-                    left_value = current_node_core_distance
-
-            if left_value < right_value:
-                current_distances[j] = left_value
-                current_sources[j] = left_source
-                if left_value < new_distance:
-                    new_distance = left_value
-                    source_node = left_source
+            if mutual_reachability_distance < next_node_min_reach:
+                min_reachability[j] = mutual_reachability_distance
+                current_sources[j] = current_node
+                if mutual_reachability_distance < new_reachability:
+                    new_reachability = mutual_reachability_distance
+                    source_node = current_node
                     new_node = j
             else:
-                if right_value < new_distance:
-                    new_distance = right_value
-                    source_node = right_source
+                if next_node_min_reach < new_reachability:
+                    new_reachability = next_node_min_reach
+                    source_node = next_node_source
                     new_node = j
 
-        result[i - 1, 0] = <double> source_node
-        result[i - 1, 1] = <double> new_node
-        result[i - 1, 2] = new_distance
+        mst[i - 1].current_node = source_node
+        mst[i - 1].next_node = new_node
+        mst[i - 1].distance = new_reachability
         current_node = new_node
 
-    return result_arr
+    return mst
 
 @cython.wraparound(True)
```

Comment:
IIRC, it is easier to maintain `wraparound` activated even if it's costly?

Reply:
Ah this was necessary because of how it was set up originally (failed otherwise) but now is no longer needed -- I forgot to clean this up. Good catch!